articles/moricho.re

```diff
@@ -0,0 +1,291 @@
+      HostID:      os.Getgid(),
+      Size:        1,
+    },
+  },
+}
+//}
+
+ここでは、ホストのユーザー名前空間と新たに分離したユーザー名前空間におけるUID/GIDのマッピングを行っています。
+なぜこうするのかというと、単にユーザー名前空間を分離しただけでは起動後のプロセス内でユーザー/グループがnobody/nogroupとなってしまうからです。
+新しいユーザー名前空間で実行されるプロセスのUID/GIDを設定するためには、@<code>{/proc/[pid]/uid_map}と@<code>{/proc/[pid]/gid_map}に対して書き込みを行います。
+Goでは@<code>{syscall.SysProcAttr}に@<code>{UidMappings}と@<code>{GidMappings}を設定することでこれをやってくれます。
+上の例ではrootユーザーとして新たなプロセスを実行しています。
+
+=== 2.ファイルシステムの隔離
+前項までは、マウント名前空間（CLONE_NEWNSフラッグで指定したもの）含む各名前空間を分離したプロセスを起動するところまでやりました。
+前項のスクリプトを実行し、起動したプロセスに入った状態でプロセス内で何がマウントされているか見てましょう。
+
+//list[mount1][実行結果][]{
+-[shoten]- # cat /proc/mounts
+/dev/xvda1 / ext4 rw,relatime,discard,data=ordered 0 0
+udev /dev devtmpfs rw,nosuid,relatime,size=491524k,nr_inodes=122881,mode=755 0 0
+devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
+...
+//}
+
+マウント空間を分離したはずなのに、ホストでマウントされている多くのマウントの情報を見ることができてしまいます。
+@<href>{http://man7.org/linux/man-pages/man7/mount_namespaces.7.html, mount_namespaces（7）}を見ると、それがなぜだかわかります。
+CLONE_NEWNSフラッグ付きで@<code>{clone()}が呼ばれた場合、呼び出し元のマウントポイントのリストが新たなプロセスのそれにコピーされる仕様になっています。
+これでは、コンテナからホストの情報が見えてしまっているためよくありません。
+そこで登場するのが@<b>{pivot_root}です。
+
+pivot_rootについて説明するにあたって、まずはファイルシステムについておさらいしましょう。
+
+==== ファイルシステムとは
+ファイルシステムとは、@<b>{ブロックデバイスのデータを構造的に扱う仕組み}のことです。ブロックデバイスはHDDやSSDなどを指します。
+ブロックデバイスのデータを人が直接扱うのは大変ですが、ファイルシステムがそれらを@<b>{ファイル}や@<b>{ディレクトリ}という形で扱うことを可能にしています。
+またブロックデバイスのデータを、システム上のディレクトリツリーに対応させることを@<b>{マウント}といいます。
+
+==== pivot_root
+ファイルシステムについて理解したところで、続いてpivot_rootについて説明していきます。
+@<b>{pivot_root}とは、プロセスのルートファイルシステムを変更するLinuxの機能です。
+@<b>{pivot_root}は引数として@<code>{new_root}と@<code>{put_old}を取ります。
+呼び出し元のプロセスのルートファイルシステムを@<code>{put_old}ディレクトリに移動させ、@<code>{new_root}を呼び出し元のプロセスの新しいルートファイルシステムにします。
+
+また@<b>{pivot_root}には、new_rootとput_oldに関して制約があります。
+
+ * ディレクトリでなければならない
+ * 現在のrootと同じファイルシステムにあってはならない
+ * put_oldはnew_rootの下になければならない
+ * ほかのファイルシステムがput_oldにマウントされていてはならない
+
+の４つです。
+
+これらを考慮して@<code>{pivot_root}を実装しましょう。
+
+//list[mount2][pivot_rootの実装][go]{
+func pivotRoot(newroot string) error {
+	putold := filepath.Join(newroot, "/oldrootfs")
+
+	// pivot_rootの条件を満たすために、新たなrootで自分自身をバインドマウント
+	if err := syscall.Mount(newroot, newroot, "", syscall.MS_BIND|syscall.MS_REC, ""); err != nil {
+		return err
+	}
+
+	if err := os.MkdirAll(putold, 0700); err != nil {
+		return err
+	}
+
+	if err := syscall.PivotRoot(newroot, putold); err != nil {
+		return err
+	}
+
+	if err := os.Chdir("/"); err != nil {
+		return err
+	}
+
+	putold = "/oldrootfs"
+	if err := syscall.Unmount(putold, syscall.MNT_DETACH); err != nil {
+		return err
+	}
+
+	if err := os.RemoveAll(putold); err != nil {
+		return err
+	}
+
+	return nil
+}
+//}
+
+流れは次のようになっています。
+
+ 1. new_rootでnew_root自身をバインドマウント（ここはあまり本質ではありません）
+ 2. pivot_rootを実行
+ 3. 不要になった以前のルートファイルシステムをアンマウント、そしてディレクトリを削除
+
+まず「new_rootとput_oldは現在のrootと同じファイルシステムにあってはならない」を満たすために、new_rootを新たなマウントポイントとしてnew_root自身でバインドマウントしています。
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Line 272 sentence length(104) exceeds the maximum sentence length of 100.
Over 4 characters. (ja-technical-writing/sentence-length)